### ISSUE TITLE
Numaplane Controller ConfigMap

### ISSUE THREAD

**Issue by juliev0** [numaproj-labs/numaplane#79]
Now that we are starting to have more configurability, we can employ a `ConfigMap` for the Numaplane Controller to read.  

We can have a `Config` struct similar to [this GlobalConfig struct in numaflow](https://github.com/numaproj/numaflow/blob/main/pkg/reconciler/config.go#L29), which is read in by a method like [this](https://github.com/numaproj/numaflow/blob/main/pkg/reconciler/config.go#L127). This file could go in our `internal/controller` directory. 

We can replace this code which looks for the CLUSTER_NAME in an environment variable [here](https://github.com/numaproj-labs/numaplane/blob/main/internal/controller/gitsync_controller.go#L59) with code which instead creates one of those `Config` structs and loads it [here](https://github.com/numaproj-labs/numaplane/blob/main/internal/controller/gitsync_controller.go#L59). The `*Config` could in fact be a member of the `GitSyncReconciler`.

This ConfigMap will include:
- cluster name
- time interval for watching repos ([this](https://github.com/numaproj-labs/numaplane/issues/77) issue)
- a mapping of RepositoryPathPrefix (e.g. "https://github.intuit.com/") to the credentials (i.e. a reference to the Kubernetes Secrets/keys within those Secrets (similar to [this](https://github.com/numaproj/numaflow/blob/main/pkg/apis/numaflow/v1alpha1/redis_buffer_service.go#L50)))

See the Numaflow Controller ConfigMap for what [theirs](https://github.com/numaproj/numaflow/blob/main/config/base/controller-manager/numaflow-controller-config.yaml) looks like as a reference. You can create one locally for your testing which has all of the fields you want to use. 

In addition, we need to add a default one to our [config/manager directory](https://github.com/numaproj-labs/numaplane/tree/main/config/manager) and incorporate into the [kustomization.yaml](https://github.com/numaproj-labs/numaplane/blob/main/config/manager/kustomization.yaml) file. You can leave the mapping of RepositoryPathPrefix to credentials empty. Time interval can be the 3 minutes which we use now. Unfortunately, we should probably have a cluster name in there. If we take the approach that is taken currently for adding the environment variable to the yaml [here](https://github.com/numaproj-labs/numaplane/blob/main/config/manager/manager.yaml#L75), then we would remove the environment variable from there, and instead would add that same "CLUSTER_NAME_VALUE" default to the ConfigMap instead, where presumably it will get set when we run `make deploy` [here](https://github.com/numaproj-labs/numaplane/blob/main/Makefile#L149).

**Comment by shubhamdixit863** [comment]
This method- ```
func LoadConfig(onErrorReloading func(error)) (*GlobalConfig, error) {
	v := viper.New()
	v.SetConfigName("controller-config")
	v.SetConfigType("yaml")
	v.AddConfigPath("/etc/numaflow")
	err := v.ReadInConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to load configuration file. %w", err)
	}
	r := &GlobalConfig{}
	err = v.Unmarshal(r)
	if err != nil {
		return nil, fmt.Errorf("failed unmarshal configuration file. %w", err)
	}
	v.WatchConfig()
	v.OnConfigChange(func(e fsnotify.Event) {
		err = v.Unmarshal(r)
		if err != nil {
			onErrorReloading(err)
		}
	})
	return r, nil
}
```
from numaflow ,loads the config from yaml and parses it to the Config  struct directly. Do we need similar approach or do we need a process to fetch the configmap from kubernetes (etcd) storage using the  k8 client api ? .As when we do kubectl apply for configmap it gets stored in etcd in cluster and available cluster wide . Please clarify @juliev0

**Comment by juliev0** [comment]
> 
> from numaflow ,loads the config from yaml and parses it to the Config struct directly. Do we need similar approach or do we need a process to fetch the configmap from kubernetes (etcd) storage using the k8 client api ? .As when we do kubectl apply for configmap it gets stored in etcd in cluster and available cluster wide . Please clarify @juliev0

You're right. I didn't look at it closely and thought it was getting it from the K8S API. In reality, it appears to be volume-mounted onto the Pod ([here](https://github.com/numaproj/numaflow/blob/main/config/base/controller-manager/controller-manager-deployment.yaml#L79) and [here](https://github.com/numaproj/numaflow/blob/main/config/base/controller-manager/controller-manager-deployment.yaml#L56), and we can do the same thing.

**Comment by juliev0** [comment]
> @juliev0 should we pass Config to GitSyncProcessor struct too ? because i need the secretKeyRef to get the secrets for github repos from the kubernetes api or is there any other way round ?

maybe just pass in the Credentials map from the Config?

**Comment by xdevxy** [comment]
Closed by #80
